tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -3024,6 +3024,20 @@ class CallGraphWithRecurseGuard {
     }
   }
 
+  // return true if FD2 is reachable from FD1
+  bool CheckReachability(FunctionDecl *FD1, FunctionDecl *FD2) {
```

Comment:
If this is only intended to stick around for debugging purposes, we should probably wrap the function in:

```c
#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
```

So that we know it is for debugging only.

Reply:
I didn't intend for this function to only serve debugging purposes. Rather, I view this function more as an implementation of a promise that the CallGraphWithRecurseGuard class made. Though the function is unused, it implements a feature that the class should implement.

Reply:
Should we call it from an `assert` somewhere to verify the promise?

Reply:
I don't see how that verifies the promise. Are you suggesting that it should have some use somewhere? That we shouldn't define a function if it's left unused?

Reply:
In general we should not leave unreachable/unused code in the repository. First and foremost because that code (by definition) isn't tested and we shouldn't add untested code.

If the `CallGraphWithRecurseGuard`, and this function verifies the promise, we could call it from an assert in situations where the promise should be true. That would both make this code live, and act as a check on the promise we expect. Also, since the call would be in an `assert` we don't slow down release builds of DXC by running an unnecessary check.